File: Data Structures & Algorithms/trapping-rain-water/submission-0.js
class Solution {
    /**
     * @param {number[]} height
     * @return {number}
     */
    trap(height) {

        if(height.length <=0) {
            return 0;
        }

        // let n = height.length;
        let res = 0;

        // for(let i=0; i<n; i++) {

        //     // ****** Get Max left and right for current position
        //     let leftMax = height[i];
        //     let rightMax = height[i]


        //     // ****** Check for left max
        //     for(let j=0; j<i; j++) {
        //         leftMax = Math.max(leftMax , height[j])
        //     }

        //     // ****** Check for left max
        //     for(let j=i+1; j<n; j++) {
        //         rightMax = Math.max(rightMax , height[j])
        //     }

        //     console.log("The left max" , leftMax)
        //     console.log("The right max" , rightMax)

        //     res += Math.min(leftMax , rightMax) - height[i]

        //      console.log("The res max" , res)

        // }


        // **** TWO Pointer

        let l = 0
        let r = height.length -1

        let leftMax = height[l]
        let rightMax = height[r]

        while(l < r) {

            if(leftMax < rightMax) {
                l++;
                leftMax = Math.max(leftMax , height[l])
                res += leftMax - height[l]
            } else {
                r--
                rightMax = Math.max(rightMax , height[r])
                res += rightMax - height[r]
            }

        }

        return res

    }
}

File: Data Structures & Algorithms/replace-elements-with-greatest-element-on-right-side/submission-0.java
class Solution {
    public int[] replaceElements(int[] arr) {
        // for (int i = 0; i < arr.length - 1; i++) {
        //     int max = arr[i + 1];

        //     for (int j = i + 1; j < arr.length; j++) {
        //         if (arr[j] > max) {
        //             max = arr[j];
        //         }
        //     }

        //     arr[i] = max;
        // }

        // arr[arr.length - 1] = -1;

        int rightMax = -1;
        int n = arr.length;
        int[] ans = new int[n];

        for(int i = n-1; i>= 0; i--){
            ans[i] = rightMax;
            rightMax = Math.max(rightMax , arr[i]);
        }

        return ans;
    }
}